devdoc/channel.md

```diff
@@ -0,0 +1,278 @@
+    CHANNEL_RESULT_INVALID_ARGS, \
+    CHANNEL_RESULT_ERROR
+
+MU_DEFINE_ENUM(CHANNEL_RESULT, CHANNEL_RESULT_VALUES);
+
+#define CHANNEL_CALLBACK_RESULT_VALUES \
+    CHANNEL_CALLBACK_RESULT_OK, \
+    CHANNEL_CALLBACK_RESULT_CANCELLED, \
+    CHANNEL_CALLBACK_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(CHANNEL_CALLBACK_RESULT, CHANNEL_CALLBACK_RESULT_VALUES);
+
+typedef void(*PULL_CALLBACK)(void* pull_context, CHANNEL_CALLBACK_RESULT result, THANDLE(RC_PTR) data);
+typedef void(*PUSH_CALLBACK)(void* push_context, CHANNEL_CALLBACK_RESULT result);
+typedef struct CHANNEL_TAG CHANNEL;
+
+#include "umock_c/umock_c_prod.h"
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+THANDLE_TYPE_DECLARE(CHANNEL);
+
+    MOCKABLE_FUNCTION(, THANDLE(CHANNEL), channel_create, THANDLE(THREADPOOL), threadpool);
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_pull, THANDLE(CHANNEL), channel, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull);
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_push, THANDLE(CHANNEL), channel, THANDLE(RC_PTR), data, PUSH_CALLBACK, push_callback, void*, push_context, THANDLE(ASYNC_OP)*, out_op_push);
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+```
+
+### channel_create
+```c
+    MOCKABLE_FUNCTION(, THANDLE(CHANNEL), channel_create, THANDLE(THREADPOOL), threadpool);
+```
+
+`channel_create` creates the channel and returns it.
+
+**SRS_CHANNEL_43_077: [** If `threadpool` is `NULL`, `channel_create` shall fail and return `NULL`. **]**
+
+**SRS_CHANNEL_43_001: [** `channel_create` shall create a `CHANNEL` object by calling `THANDLE_MALLOC` with `channel_dispose` as `dispose`. **]**
+
+**SRS_CHANNEL_43_078: [** `channel_create` shall create a `CHANNEL_INTERNAL` object by calling `THANDLE_MALLOC` with `channel_internal_dispose` as `dispose`.**]**
+
+**SRS_CHANNEL_43_079: [** `channel_create` shall store the created `THANDLE(CHANNEL_INTERNAL)` in the `THANDLE(CHANNEL)`. **]**
+
+**SRS_CHANNEL_43_080: [** `channel_create` shall initialize the `CHANNEL_INTERNAL` with the given `threadpool`. **]**
+
+**SRS_CHANNEL_43_098: [** `channel_create` shall call `srw_lock_create`. **]**
+
+**SRS_CHANNEL_43_084: [** `channel_create` shall call `DList_InitializeListHead`. **]**
+
+**SRS_CHANNEL_43_086: [** `channel_create` shall succeed and return the created `THANDLE(CHANNEL)`. **]**
+
+**SRS_CHANNEL_43_002: [** If there are any failures, `channel_create` shall fail and return `NULL`. **]**
+
+
+### channel_dispose
+```c
+    static void channel_dispose(CHANNEL* channel);
+```
+
+`channel_dispose` disposes the given `channel` and schedules all pending operations to be abandoned.
+
+**SRS_CHANNEL_43_094: [** `channel_dispose` shall call `srw_lock_acquire_exclusive`. **]**
+
+**SRS_CHANNEL_43_095: [** `channel_dispose` shall iterate over the list of pending operations and do the following: **]**
+
+ - **SRS_CHANNEL_43_096: [** set the `result` of the `operation` to `CHANNEL_CALLBACK_RESULT_ABANDONED`. **]**
```

Comment:
>CHANNEL_CALLBACK_RESULT_ABANDONED [](http://example.com/codeflow?start=69&length=33)

Will the operations that were given to the threadpool equally set to `ABANDONED`?

Reply:
If they are in the threadpool, channel module does not hold a reference to them anymore, so they cannot be set to abandoned.